### DISCUSSION TITLE
Access Pico W pins

### DISCUSSION THREAD

**Discussion by NuclearPhoenixx** [earlephilhower/arduino-pico#835] [Q&A] [answered]
Hi, I just had a look at the datasheet for the Pico W and was wondering how to access these pins:

![Screenshot 2022-09-04 at 11-57-18 Raspberry Pi Pico W Datasheet - pico-w-datasheet pdf](https://user-images.githubusercontent.com/17517687/188307913-1b617df3-96c8-4224-b711-ab7cade7f859.png)

These are different to the digital pins 0, 1 and 2 and `WL_GPIO0` or `WL0` does not compile correctly.

**Comment by earlephilhower** [comment, marked as answer]
My understanding, after going thru the CYW43 code bringing up WiFi is that WL0 == WL_GPIO0, and the other 2 pins are used as part of the WiFi PA and not general purpose.  They are not accessible from the Pico except thru the CYW43 chip + binary blob and driver.

You might get a better answer at the Raspberry Pi Pico forums to a HW question like this:  https://forums.raspberrypi.com

**Reply by earlephilhower** [reply]
(and `LED_BUILTIN` is WL0 and is driven thru API calls to the CYW43 driver).

**Reply by NuclearPhoenixx** [reply]
Alright, thank you! Thought there was a way to use it directly through arduino-pico, but doesn't seem like it.

**Reply by wohali** [reply]
Based on [this thread](https://forums.raspberrypi.com/viewtopic.php?p=2047656) the "official" answer is:

`cyw43_arch_gpio_put(1, 1);`

Given https://github.com/earlephilhower/arduino-pico/blob/master/variants/rpipicow/picow_digital.cpp#L35-L41 , with this core the following should work:

`digitalWrite(32+1, 1);`

One option for @earlephilhower might be to add the following to `/variants/rpipicow/pins_arduino.h`:

```C
#define PIN_WL_GPIO0 (32u)
#define PIN_WL_GPIO1 (33u)
#define PIN_WL_GPIO2 (34u)
```

or, more along the lines of `/variants/generic/common.h`:

```C
static const uint8_t WL_GPIO0 = (32u);
static const uint8_t WL_GPIO1 = (33u);
static const uint8_t WL_GPIO2 = (34u);
```

but given this is colouring outside the intended lines of this core, you could just add these to your own project.

FYI: I came here looking for a way to control the W's buck-boost SMPS mode, but after reading [Mark Omo's ADC writeup](https://pico-adc.markomo.me/) and analysis of power supply noise, I decided against it (and use an external Vref source).

**Reply by NuclearPhoenixx** [reply]
Very interesting, thanks for looking that up @wohali!

> FYI: I came here looking for a way to control the W's buck-boost SMPS mode, but after reading [Mark Omo's ADC writeup](https://pico-adc.markomo.me/) and analysis of power supply noise, I decided against it (and use an external Vref source).

I absolutely love Mark Omo's ADC writeup. But yes, if you can spare a $1 LM4040, I would highly recommend it to anyone using the Pico ADC. The output precision on the Pico SMPS is **not** great. Not even speaking of the noise.

**Reply by donmsmall** [reply]
Using the following definitions in my code has worked perfectly until recently.  I'm using PIN_WL_GPIO2 for determining if I have VBUS applied (using USB) or not.  If my system is battery powered (not USB) then it needs to run with watchdog timers enabled and no need to offer a test menu because without a USB connection I won't have a serial port available.

```c++
#define PIN_WL_GPIO0 (32u)    
#define PIN_WL_GPIO1 (33u)
#define PIN_WL_GPIO2 (34u)
``` 
I've recently updated from core 3.9.3 to 4.4.0 and now checking PIN_WL_GPIO2 to determine if the USB serial port is available doesn't work anymore.  What I have found is that some of the functionality previously found by @wohali in  https://github.com/earlephilhower/arduino-pico/blob/master/variants/rpipicow/picow_digital.cpp#L35-L41 has now moved to cyw43_wrappers.cpp.  It contains the following code snippet:
```c++
extern "C" void cyw43_digitalWrite(pin_size_t pin, PinStatus val) {
    if (!__isPicoW && (pin == PIN_LED)) {
        pin = 25;  // Silently swap in the Pico's LED
    }
    if (pin < 64) {
        __digitalWrite(pin, val);
    } else {
        cyw43_arch_gpio_put(pin - 64, val == HIGH ? 1 : 0);
    }
}
```
where the similar code from picow_digital.cpp looks like this
```c++
extern "C" void __digitalWrite(pin_size_t pin, PinStatus val) {
    if (!__isPicoW && (pin == PIN_LED)) {
        pin = 25;  // Silently swap in the Pico's LED
    }
    if (pin < 32) {
        __digitalWrite(pin, val);
    } else {
        cyw43_arch_gpio_put(pin - 32, val == HIGH ? 1 : 0);
    }
}
```
Apparently, when moving to the new rp2350 architecture, there are more gpio pins available and thus (pin < 32) needed to be changed to (pin < 64).  I also see that PIN_LED = 32 for 3.9.3 and is PIN_LED = 64 for 4.4.0.  To fix my code and make it back compatible with previous rp2040 core versions, I'm changing my definitions to the following:

```c++
#define PIN_WL_GPIO0 (PIN_LED)    
#define PIN_WL_GPIO1 (PIN_LED+1u)
#define PIN_WL_GPIO2 (PIN_LED+2u)
```

**Comment by kendo55** [comment]
here's an example

````

#include "pico/cyw43_arch.h"
#include "boards/pico_w.h"

void blinky_WL_led(uint16_t lamda_ms, uint16_t flash_ms){
//-------------------------------------------------------
  static uint32_t alt_ms = 0;
  static uint8_t led_on = 0;
 
  if ( (millis() - alt_ms) >= ( led_on ? flash_ms : lamda_ms - flash_ms )){    
    alt_ms = millis();
    led_on = ! led_on;
    cyw43_arch_gpio_put(CYW43_WL_GPIO_LED_PIN, led_on);
  }
}

void loop() {
//----------------------
.
.
blinky_WL_led(1000, 50);
.
}
````

**Reply by earlephilhower** [reply]
You could do this, but it would be wasted effort.  So I'd really recommend not bothering.

The LED pin is the only one that's generally user controlled on the CYW43 chip, and it works with `digitalWrite(LED_DEFAULT, xxx)` just fine on the Pico W.

**Comment by kendo55** [comment]
Sorry ...

with digitalWrite(LED_DEFAULT, xxx), I received that error:

'LED_DEFAULT' was not declared in this scope

I searched  'LED_DEFAULT' in ...packages\rp2040\hardware\rp2040\1.9.8

but it's not #defined anywhere

**Reply by earlephilhower** [reply]
1.9.8 is  old and may not even have PicoW support IIRC.  I suggest you update.  We're in the 2.7.xs now.

As for that define, it should be `LED_DEFAULT` as in the Blink.ino example, my typo.

**Comment by kendo55** [comment]
My Version is the 2.7.0 sorry!

AND it should be  digitalWrite(LED_BUILTIN, xxx)

then it works. 

thanks!

**Comment by jorgechacblogspot** [comment]


**Reply by Andy2No** [reply]
@jorgechacblogspot are you using LED_BUILTIN to refer to the pin?

**Comment by Andy2No** [comment]
@jorgechacblogspot  try LED_BUILTIN instead.

**Reply by earlephilhower** [reply]
The full error message explains exactly the problem and solution.

static_assert(0, "stdio_init_all is not supported or needed. Either use Serial.printf() or set the debug port in the IDE to Serial/1/2 and use printf().  See https://github.com/earlephilhower/arduino-pico/issues/1433#issuecomment-1540354673 and https://github.com/earlephilhower/arduino-pico/issues/1433#issuecomment-1546783109")`

In any case don't call the cyw43* calls directly, not use the stdio_init* either since both of those bits of HW are controlled by the core itself.  Calling them in your own app will cause much sadness when things get confused.

**Comment by Andy2No** [comment]
@jorgechacblogspot Does this work? (with no include files):

```
/*
  Based on Blink
  This version does a heart beat double flash, once per second.
 */
 
int led = LED_BUILTIN; // Pin 13 has an LED connected on most Arduino boards.

void setup() {
  pinMode(led, OUTPUT);     
}

// the loop routine runs over and over again forever:
void loop() {
  digitalWrite(led, HIGH);
  delay(100);               // wait for 0.1 second
  digitalWrite(led, LOW); 
  delay(100);               // wait for 0.1 second
  digitalWrite(led, HIGH);
  delay(200);               // wait for 0.2 second
  digitalWrite(led, LOW);
  delay(600);               // wait for 0.6 second
}
```

**Reply by Andy2No** [reply]
@jorgechacblogspot but did you try it?

There's supposed to be some trickery where using the LED_BUILTIN definition behaves the same way on a Pico W as on a Pico.  The digitalWrite() gets translated to code to use the LED, as I understand it.  I don't own a Pico W, but as far as I'm aware, that's supposed to work.

The pin 13 comment is just left over from running that sketch on an Uno, or similar.

**Reply by earlephilhower** [reply]
That is taken care of behind the scenes.  Just use `LED_BUILTIN` and `digitalWrite` and everything "just works."  Make sure you have selected the Pico W board from the menus, of course.

**Reply by earlephilhower** [reply]
I still think you're missing the main idea.  If you want to write raw SDK code, that's fine but don't try doing that with the Arduino core.  Doing manual CYW43 inits and manually adding `printf` prototypes is not going to end well.  Just the simple code @Andy2No posted should run like a champ, or even look at the `blink.ino` example from Arduino itself.

**Reply by savejeff** [reply]
Just for people finding this: 
the current define is PIN_LED not LED_BUILTIN
